### ISSUE TITLE
JS error interacting with SimplePress

### ISSUE THREAD

**Issue by riverofjanuary** [MinnPost/object-sync-for-salesforce#156]
When we have SimplePress installed, we get a $ library error on the main page:

![image](https://user-images.githubusercontent.com/1587594/37529110-a60b92f6-290c-11e8-9bf0-41dd002f7ecf.png)

**Comment by jonathanstegall** [comment]
@riverofjanuary I'm not familiar with SimplePress. It doesn't seem to be in the WP plugin directory.

But in any case, I made a few JavaScript changes recently that haven't officially been released. I wonder if you'd be willing to test them?

If you're familiar with Git, you could install them from the current master branch of this repository. If not, try replacing the file at `object-sync-for-salesforce/assets/js/object-sync-for-salesforce-admin.min.js` with the following code:

```
"use strict";!function($){function salesforce_object_fields(){0===$(".salesforce_record_types_allowed > *").length&&$(".salesforce_record_types_allowed").hide(),0===$(".salesforce_record_type_default > *").length&&$(".salesforce_record_type_default").hide(),0===$(".pull_trigger_field > *").length&&$(".pull_trigger_field").hide();var delay=function(){var timer=0;return function(callback,ms){clearTimeout(timer),timer=setTimeout(callback,ms)}}();$("#salesforce_object").on("change",function(el){var that=this;delay(function(){var data={action:"get_salesforce_object_description",include:["fields","recordTypeInfos"],field_type:"datetime",salesforce_object:that.value};$.post(ajaxurl,data,function(response){var record_types_allowed_markup="",record_type_default_markup="",date_markup="";$(response.data.recordTypeInfos).length>0&&(record_types_allowed_markup+='<label for="salesforce_record_types_allowed">Allowed Record Types:</label><div class="checkboxes">',$.each(response.data.recordTypeInfos,function(index,value){record_types_allowed_markup+='<label><input type="checkbox" class="form-checkbox" value="'+index+'" name="salesforce_record_types_allowed['+index+']" id="salesforce_record_types_allowed-'+index+'"> '+value+"</label>"}),record_types_allowed_markup+="</div>",record_type_default_markup+='<label for="salesforce_record_type_default">Default Record Type:</label>',record_type_default_markup+='<select name="salesforce_record_type_default" id="salesforce_record_type_default"><option value="">- Select record type -</option>',$.each(response.data.recordTypeInfos,function(index,value){record_type_default_markup+='<option value="'+index+'">'+value+"</option>"})),$(".salesforce_record_types_allowed").html(record_types_allowed_markup),$(".salesforce_record_type_default").html(record_type_default_markup),$(response.data.fields).length>0&&(date_markup+='<label for="pull_trigger_field">Date field to trigger pull:</label>',date_markup+='<select name="pull_trigger_field" id="pull_trigger_field"><option value="">- Select date field -</option>',$.each(response.data.fields,function(index,value){date_markup+='<option value="'+value.name+'">'+value.label+"</option>"}),date_markup+="</select>",date_markup+='<p class="description">These are date fields that can cause WordPress to pull an update from Salesforce, according to the <code>salesforce_pull</code> class.</p>'),$(".pull_trigger_field").html(date_markup),""!==record_types_allowed_markup?$(".salesforce_record_types_allowed").show():$(".salesforce_record_types_allowed").hide(),""!==record_type_default_markup?$(".salesforce_record_type_default").show():$(".salesforce_record_type_default").hide(),""!==date_markup?$(".pull_trigger_field").show():$(".pull_trigger_field").hide()})},1e3)})}function add_field_mapping_row(){$("#add-field-mapping").click(function(){$(this).text("Add another field mapping");var salesforce_object=$("#salesforce_object").val(),wordpress_object=$("#wordpress_object").val();return""!==wordpress_object&&""!==salesforce_object?(fieldmap_fields(wordpress_object,salesforce_object,Math.floor(Date.now()/1e3)),$(this).parent().find(".missing-object").remove()):$(this).parent().prepend('<div class="error missing-object"><span>You have to pick a WordPress object and a Salesforce object to add field mapping.</span></div>'),!1})}function fieldmap_fields(wordpress_object,salesforce_object,row_key){var data={action:"get_wp_sf_object_fields",wordpress_object:wordpress_object,salesforce_object:salesforce_object};$.post(ajaxurl,data,function(response){var wordpress="";wordpress+='<select name="wordpress_field['+row_key+']" id="wordpress_field-'+row_key+'">',wordpress+='<option value="">- Select WordPress field -</option>',$.each(response.data.wordpress,function(index,value){wordpress+='<option value="'+value.key+'">'+value.key+"</option>"}),wordpress+="</select>";var salesforce="";salesforce+='<select name="salesforce_field['+row_key+']" id="salesforce_field-'+row_key+'">',salesforce+='<option value="">- Select Salesforce field -</option>',$.each(response.data.salesforce,function(index,value){salesforce+='<option value="'+value.name+'">'+value.label+"</option>"});var markup='<tr><td class="column-wordpress_field">'+wordpress+'</td><td class="column-salesforce_field">'+(salesforce+="</select>")+'</td><td class="column-is_prematch"><input type="checkbox" name="is_prematch['+row_key+']" id="is_prematch-'+row_key+'" value="1" /><td class="column-is_key"><input type="checkbox" name="is_key['+row_key+']" id="is_key-'+row_key+'" value="1" /></td><td class="column-direction"><div class="radios"><label><input type="radio" value="sf_wp" name="direction['+row_key+']" id="direction-'+row_key+'-sf-wp">  Salesforce to WordPress</label><label><input type="radio" value="wp_sf" name="direction['+row_key+']" id="direction-'+row_key+'-wp-sf">  WordPress to Salesforce</label><label><input type="radio" value="sync" name="direction['+row_key+']" id="direction-'+row_key+'-sync" checked>  Sync</label></div></td><td class="column-is_delete"><input type="checkbox" name="is_delete['+row_key+']" id="is_delete-'+row_key+'" value="1" /></td></tr>';$("table.fields tbody").append(markup)})}function push_and_pull_objects(){$(".salesforce_user_ajax_message").hide(),$("#wordpress_object_ajax").length>0&&$(".push_to_salesforce_button").on("click",function(){var data={action:"push_to_salesforce",wordpress_object:$("#wordpress_object_ajax").val(),wordpress_id:$("#wordpress_id_ajax").val()};return $.post(ajaxurl,data,function(response){!0===response.success&&(update_salesforce_user_summary(),$(".salesforce_user_ajax_message").width($(".mapped-salesforce-user").width()-27),$(".salesforce_user_ajax_message").html("<p>This object has been pushed to Salesforce.</p>").fadeIn().delay(4e3).fadeOut())}),!1}),$(".pull_from_salesforce_button").on("click",function(){var data={action:"pull_from_salesforce",salesforce_id:$("#salesforce_id_ajax").val(),wordpress_object:$("#wordpress_object_ajax").val()};return $.post(ajaxurl,data,function(response){!0===response.success&&(update_salesforce_user_summary(),$(".salesforce_user_ajax_message").width($(".mapped-salesforce-user").width()-27),$(".salesforce_user_ajax_message").html("<p>This object has been pulled from Salesforce.</p>").fadeIn().delay(4e3).fadeOut())}),!1})}function update_salesforce_user_summary(){var data={action:"refresh_mapped_data",mapping_id:$("#mapping_id_ajax").val()};$.post(ajaxurl,data,function(response){!0===response.success&&($("td.last_sync_message").text(response.data.last_sync_message),$("td.last_sync_action").text(response.data.last_sync_action),$("td.last_sync_status").text(response.data.last_sync_status),$("td.last_sync").text(response.data.last_sync),"1"===response.data.last_sync_status&&$("td.last_sync_status").text("success"))})}function clear_sfwp_cache_link(){$("#clear-sfwp-cache").click(function(){var data={action:"clear_sfwp_cache"},that=$(this);return $.post(ajaxurl,data,function(response){!0===response.success&&!0===response.data.success&&that.parent().html(response.data.message).fadeIn()}),!1})}$(document).on("click",".column-is_prematch input",function(){$(".column-is_prematch input").not(this).prop("checked",!1)}),$(document).on("click",".column-is_key input",function(){$(".column-is_key input").not(this).prop("checked",!1)}),$(document).ready(function(){var timeout;$("#wordpress_object, #salesforce_object").on("change",function(){clearTimeout(timeout),timeout=setTimeout(function(){$("table.fields tbody tr").fadeOut(),$("table.fields tbody tr").remove()},1e3)}),$(document).ajaxStart(function(){$(".spinner").addClass("is-active")}).ajaxStop(function(){$(".spinner").removeClass("is-active")}),salesforce_object_fields(),add_field_mapping_row(),push_and_pull_objects(),clear_sfwp_cache_link()})}(jQuery);
```

You would probably need to clear your browser cache if you replace it. You could also change line 407 in the main object-sync-for-salesforce.php file to the following, and it will make sure you load a newer version.

```
wp_enqueue_script( $this->slug . '-admin', plugins_url( 'assets/js/object-sync-for-salesforce-admin.min.js', __FILE__ ), array( 'jquery' ), filemtime( plugin_dir_path( __FILE__ ) . 'assets/js/object-sync-for-salesforce-admin.js' ), true );
```

Anyway if this fixes the issue you're having, I'll release a plugin update to put the fix out there.

**Comment by riverofjanuary** [comment]
Jonathan - it worked! Fantastic fix, thank you. Did you just check if $ had been instantiated? Whatever you did, thank you. SimplePress is an antiquated forum plugin but one we're forced to use for a current client. Thank you.

**Comment by jonathanstegall** [comment]
Here's the commit that addressed this: d35a6a899b75aa551acaf0562f5304b379b3fa73. It just wraps the $ in an enclosure.

I had noticed that some users would report that this issue was happening. It does seem like it was mostly older plugins, but enough to be troubling.

Anyway I'll release this as a new version shortly.

**Comment by jonathanstegall** [comment]
Version 1.3.2 has been released, and it contains this fix. I'll mark this issue closed, unless it arises again.

Thanks for the flag to get it out there.
